dcicutils/qa_utils.py

```diff
@@ -678,15 +682,487 @@ def Session(self, **kwargs):  # noQA - This name was chosen by AWS, so please do
+
+    _SHARED_DATA_MARKER = "_IAM_SHARED_DATA_MARKER"
+
+    def __init__(self, boto3=None) -> None:
+        self.boto3 = boto3 or MockBoto3()
+
+    class _UserAccessKeyPair:
+        def __init__(self) -> None:
+            self._id = str(uuid.uuid4())
+            self._secret = str(uuid.uuid4())
+            self._create_date = datetime.datetime.now()
+        @property
+        def id(self) -> str:
+            return self._id
+        @property
+        def secret(self) -> str:
+            return self._secret
+        def get(self, key: str) -> Any:
+            if key == "AccessKeyId":
+                return self._id
+            elif key == "CreateDate":
+                return self._create_date
+            return None
+        def __getitem__(self, key: str) -> Any:
+            return self.get(key)
+
+    class _UserAccessKeyPairCollection:
+        def __init__(self) -> None:
+            self._access_key_pairs = []
+        def add(self, access_key_pair: MockBoto3Iam._UserAccessKeyPair) -> None:
+            self._access_key_pairs.append(access_key_pair)
+        def get(self, key: str) -> Any:
+            if key == "AccessKeyMetadata":
+                return self._access_key_pairs
+            return None
+        def __getitem__(self, key: str) -> Any:
+            return self.get(key)
+
+    class _User:
+        def __init__(self, name: str) -> None:
+            self._name = name
+            self._access_keys = MockBoto3Iam._UserAccessKeyPairCollection()
+        @property
+        def name(self) -> str:
+            return self._name
+        def create_access_key_pair(self) -> MockBoto3Iam._UserAccessKeyPair:
+            access_key_pair = MockBoto3Iam._UserAccessKeyPair()
+            self._access_keys.add(access_key_pair)
+            return access_key_pair
+
+    class _UserCollection:
+        def __init__(self) -> None:
+            self._users = []
+        def all(self) -> list[MockBoto3Iam._User]:
+            return self._users
+
+    class _RoleCollection:
+        def __init__(self) -> None:
+            self._roles = []
+        def __getitem__(self, key: str) -> Any:
+            if key == "Roles":
+                return self._roles
+            return None
+
+    class _Role:
+        def __init__(self, arn: str) -> None:
+            self._arn = arn
+        def __getitem__(self, key: str) -> Any:
+            if key == "Arn":
+                return self._arn
+            return None
+
+    def _mocked_shared_data(self) -> dict:
+        shared_reality = self.boto3.shared_reality
+        shared_data = shared_reality.get(self._SHARED_DATA_MARKER)
+        if shared_data is None:
+            shared_data = shared_reality[self._SHARED_DATA_MARKER] = {}
+            shared_data["users"] = MockBoto3Iam._UserCollection()
+            shared_data["roles"] = MockBoto3Iam._RoleCollection()
+        return shared_data
+
+    def _mocked_users(self) -> Any:
+        return self._mocked_shared_data()["users"]
```

Comment:
Be careful about assuming shared data contains the attributes you expect. You might want `.get("users")` here? Or you might want to have a function you call to assure it's there before you do this access.